### PR TITLE
kv: reduce gRPC window size to avoid OOM

### LINKED ISSUE #2673: gRPC initial window is too large that may cause OOM
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.

Capture 10k tables with ~90G incremental data using single CDC node.

heap profile: [profile.pb.gz](https://github.com/pingcap/ticdc/files/7075353/profile.pb.gz)

| gRPC client | flamegraph |
| -- | -- |
image | image

2. What did you expect to see?

No OOM.

3. What did you see instead?

gRPC client consumes too many memory and cause OOM.

4. Versions of the cluster

v5.1.0

### PR DESCRIPTION
**Pull request by overvenus**
### What problem does this PR solve? 

Note: TiCDC in the following tests includes https://github.com/pingcap/ticdc/pull/2214.  

### Memory usage (pause sorter)

Tested with following changes

```diff
diff --git a/cdc/puller/sorter/unified_sorter.go b/cdc/puller/sorter/unified_sorter.go
index 9cd9c898..6e033429 100644
--- a/cdc/puller/sorter/unified_sorter.go
+++ b/cdc/puller/sorter/unified_sorter.go
@@ -197,9 +197,9 @@ func (s *UnifiedSorter) Run(ctx context.Context) error {
                }
        })
 
-       errg.Go(func() error {
-               return printError(runMerger(subctx, numConcurrentHeaps, heapSorterCollectCh, s.outputCh, ioCancelFunc))
-       })
+       // errg.Go(func() error {
+       //      return printError(runMerger(subctx, numConcurrentHeaps, heapSorterCollectCh, s.outputCh, ioCancelFunc))
+       // })
 
        errg.Go(func() error {
                captureAddr := util.CaptureAddrFromCtx(ctx)
```

| 64MB vs 64KB |
| -- |
| image |
| image |
| image |

The last image shows TiKV snapshot is holded for a long time, because CDC scan is flow controlled. Through, I doubt if it will actually happen as TiCDC sorter can always make progress.

## Performance

No significant difference.

| 64MB (OOM) vs 64KB |
| -- |
| image |
| image |
| image |

Close #2673 

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
Fix OOM when TiCDC captures too many regions
```

### CODE REVIEW

**Comment by amyangfei** [comment]
By the way, is there any metric exposed from gRPC API we can use to monitor the stream/connection status, such as window size, receive buffer, etc.

**Comment by overvenus** [comment]
> By the way, is there any metric exposed from gRPC API we can use to monitor the stream/connection status, such as window size, receive buffer, etc.

Maybe you are looking for channelz https://pkg.go.dev/google.golang.org/grpc/internal/channelz#ChannelInternalMetric

**Review by liuzix — APPROVED** [review]

File: cdc/kv/client.go [hunk 1]
```diff
@@ -52,10 +52,10 @@ import (
 const (
 	dialTimeout           = 10 * time.Second
 	tikvRequestMaxBackoff = 20000 // Maximum total sleep time(in ms)
-	// TODO find optimal values and test extensively before releasing
-	// The old values cause the gRPC stream to stall for some unknown reason.
-	grpcInitialWindowSize     = 1 << 26 // 64 MB The value for initial window size on a stream
-	grpcInitialConnWindowSize = 1 << 27 // 128 MB The value for initial window size on a connection
+	// TiCDC may open numerous gRPC streams,
+	// with 64KB window size, 10K streams takes about 27GB memory.
+	grpcInitialWindowSize     = 65535   // 64 KB The value for initial window size on a stream
```
Comment (3AceShowHand) [inline cdc/kv/client.go:57]: how about `1 << 16` ?

**Pull request merged**